compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp

```diff
@@ -958,5 +959,68 @@ void transform_dialect::WorkgroupSwizzleOp::getEffects(
   transform::modifiesPayload(effects);
 }
 
+//===----------------------------------------------------------------------===//
+// GpuDistributeSharedMemoryCopyOp
+//===----------------------------------------------------------------------===//
+
+DiagnosedSilenceableFailure
+transform_dialect::GpuDistributeSharedMemoryCopyOp::applyToOne(
+    transform::TransformRewriter &rewriter, func::FuncOp target,
+    transform::ApplyToEachResultList &results,
+    transform::TransformState &state) {
+
+  // Look for ops that move to/from workgroup memory and mark as copies for
+  // distribution.
+  target.walk([&](linalg::GenericOp copyOp) {
+    if (copyOp.getDpsInputOperands().size() != 1 ||
+        copyOp.getDpsInitOperands().size() != 1)
+      return;
+    auto sourceType =
+        copyOp.getDpsInputOperand(0)->get().getType().dyn_cast<MemRefType>();
+    auto destType =
+        copyOp.getDpsInitOperand(0)->get().getType().dyn_cast<MemRefType>();
+    if (!sourceType || !destType)
+      return;
+
+    Block &body = copyOp.getRegion().front();
+    if (!std::begin(body)->hasTrait<OpTrait::IsTerminator>())
+      return;
+
+    auto sourceSpace = sourceType.getMemorySpace();
+    auto destSpace = destType.getMemorySpace();
+    if ((!sourceSpace && !destSpace) || sourceSpace == destSpace)
+      return;
+
+    auto sourceGpuSpace = sourceSpace.dyn_cast_or_null<gpu::AddressSpaceAttr>();
+    auto destGpuSpace = destSpace.dyn_cast_or_null<gpu::AddressSpaceAttr>();
+    if ((!sourceGpuSpace || sourceGpuSpace.getValue() !=
+                                gpu::GPUDialect::getWorkgroupAddressSpace()) &&
+        (!destGpuSpace || destGpuSpace.getValue() !=
+                              gpu::GPUDialect::getWorkgroupAddressSpace()))
+      return;
+
+    setMarker(copyOp, getCopyToWorkgroupMemoryMarker());
+    // Insert barriers immediately after copies to workgroup memory.
+    if (destGpuSpace && destGpuSpace.getValue() ==
```

Comment:
What about copies from shared memory to global memory? Don't we want to add barriers after that as well?